AsyncDisplayKit/Details/ASRangeController.mm

```diff
@@ -159,42 +192,65 @@ - (void)updateVisibleNodeIndexPaths
     return;
   }
 
-  NSArray *indexPaths = [_delegate rangeControllerVisibleNodeIndexPaths:self];
+  NSArray *visibleNodePaths = [_delegate rangeControllerVisibleNodeIndexPaths:self];
+  NSSet *visibleNodePathsSet = [NSSet setWithArray:visibleNodePaths];
   CGSize viewportSize = [_delegate rangeControllerViewportSize:self];
 
-  if ([_layoutController shouldUpdateWorkingRangesForVisibleIndexPath:indexPaths viewportSize:viewportSize]) {
-    [_layoutController setVisibleNodeIndexPaths:indexPaths];
-    NSSet *workingRangeIndexPaths = [_layoutController workingRangeIndexPathsForScrolling:_scrollDirection viewportSize:viewportSize];
-    NSSet *visibleRangeIndexPaths = [NSSet setWithArray:indexPaths];
-
-    NSMutableSet *removedIndexPaths = _workingRangeIsValid ? [_workingRangeIndexPaths mutableCopy] : [NSMutableSet set];
-    [removedIndexPaths minusSet:workingRangeIndexPaths];
-    [removedIndexPaths minusSet:visibleRangeIndexPaths];
-    if (removedIndexPaths.count) {
-      NSArray *removedNodes = [_delegate rangeController:self nodesAtIndexPaths:[removedIndexPaths allObjects]];
-      [removedNodes enumerateObjectsUsingBlock:^(ASCellNode *node, NSUInteger idx, BOOL *stop) {
-        [self removeNodeFromWorkingRange:node];
-      }];
+  // the layout controller needs to know what the current visible indices are to calculate range offsets
+  [_layoutController setVisibleNodeIndexPaths:visibleNodePaths];
+
+  for (NSInteger i = 0; i < ASLayoutRangeCount; i++) {
+    ASLayoutRange range = (ASLayoutRange)i;
+    id rangeKey = @(range);
+
+    if ([_layoutController shouldUpdateForVisibleIndexPaths:visibleNodePaths viewportSize:viewportSize range:range]) {
+      NSSet *indexPaths = [_layoutController indexPathsForScrolling:_scrollDirection viewportSize:viewportSize range:range];
+
+      // Notify to remove indexpaths that are leftover that are not visible or included in the _layoutController calculated paths
+      NSMutableSet *removedIndexPaths = _rangeIsValid ? [[_rangeIndexPaths objectForKey:rangeKey] mutableCopy] : [NSMutableSet set];
+      [removedIndexPaths minusSet:indexPaths];
+      [removedIndexPaths minusSet:visibleNodePathsSet];
+      if (removedIndexPaths.count) {
+        NSArray *removedNodes = [_delegate rangeController:self nodesAtIndexPaths:[removedIndexPaths allObjects]];
+        [removedNodes enumerateObjectsUsingBlock:^(ASCellNode *node, NSUInteger idx, BOOL *stop) {
+          [self removeNode:node fromRange:range];
+        }];
+      }
+
+      // Notify to add indexpaths that are not currently in _rangeIndexPaths
+      NSMutableSet *addedIndexPaths = [indexPaths mutableCopy];
+      [addedIndexPaths minusSet:[_rangeIndexPaths objectForKey:rangeKey]];
+
+      // The preload range (for example) should include nodes that are visible
+      if ([self shouldRemoveVisibleNodesFromRange:range]) {
```

Comment:
why do we need to this check ? I assume the visible node will trigger the preloading directly.